Threaded build broken on Darwin 15.4

For the past week or so I have been unable to build Julia with threading enabled (`JULIA_THREADS=1`). Building with threading disabled works fine. This is the last output and error I see:
```
 clang++ -stdlib=libc++ -mmacosx-version-min=10.7 -m64 -shared -pipe -fPIC -fno-rtti -pedantic  -O3 -g -D_GNU_SOURCE -I. -I/Users/eschnett/src/julia05/src -I/Users/eschnett/src/julia05/src/flisp -I/Users/eschnett/src/julia05/src/support -I/Users/eschnett/src/julia05/usr/include -I/Users/eschnett/src/julia05/usr/include -DLIBRARY_EXPORTS -I/Users/eschnett/src/julia05/deps/valgrind -Wall -Wno-strict-aliasing -fno-omit-frame-pointer -fvisibility=hidden -fno-common -Wpointer-arith -Wundef -I/Users/eschnett/src/julia05/usr/include "-DJL_SYSTEM_IMAGE_PATH=\"../lib/julia/sys.dylib\"" ./jltypes.o ./gf.o ./ast.o ./builtins.o ./module.o ./interpreter.o ./alloc.o ./dlload.o ./sys.o ./init.o ./task.o ./array.o ./dump.o ./toplevel.o ./jl_uv.o ./jlapi.o ./signal-handling.o ./simplevector.o ./APInt-C.o ./runtime_intrinsics.o ./runtime_ccall.o ./threadgroup.o ./threading.o ./stackwalk.o ./gc.o ./codegen.o ./disasm.o ./debuginfo.o ./llvm-simdloop.o ./llvm-gcroot.o -Wl,-rpath,'@loader_path/' -o /Users/eschnett/src/julia05/usr/lib/libjulia.dylib   -Xlinker -all_load ./flisp/libflisp.a -Xlinker -all_load ./support/libsupport.a -L/Users/eschnett/src/julia05/usr/lib -L/Users/eschnett/src/julia05/usr/lib /Users/eschnett/src/julia05/usr/lib/libuv.a /Users/eschnett/src/julia05/usr/lib/libutf8proc.a  -L/Users/eschnett/src/julia05/usr/lib  -lLLVMLTO -lLLVMObjCARCOpts -lLLVMLinker -lLLVMBitWriter -lLLVMIRReader -lLLVMMIRParser -lLLVMAsmParser -lLLVMLibDriver -lLLVMOption -lLLVMDebugInfoPDB -lLLVMTableGen -lLLVMOrcJIT -lLLVMPasses -lLLVMipo -lLLVMVectorize -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMSelectionDAG -lLLVMAsmPrinter -lLLVMX86Desc -lLLVMMCDisassembler -lLLVMX86Info -lLLVMX86AsmPrinter -lLLVMX86Utils -lLLVMMCJIT -lLLVMDebugInfoDWARF -lLLVMLineEditor -lLLVMInterpreter -lLLVMExecutionEngine -lLLVMRuntimeDyld -lLLVMCodeGen -lLLVMTarget -lLLVMScalarOpts -lLLVMProfileData -lLLVMObject -lLLVMMCParser -lLLVMBitReader -lLLVMInstCombine -lLLVMInstrumentation -lLLVMTransformUtils -lLLVMipa -lLLVMMC -lLLVMAnalysis -lLLVMCore -lLLVMSupport -L/Users/eschnett/src/julia05/usr/lib  -lz -lm  -ldl -Wl,-w -framework CoreFoundation -framework CoreServices /Users/eschnett/src/julia05/usr/lib/libosxunwind.a 
ld: section __DATA/__thread_bss extends beyond end of file, file './threading.o' for architecture x86_64
```

This is on Darwin 15.4.0 (OS X "El Capitan" 10.11.4). I'm building with gcc 5.3 and clang 3.7.1 that were installed via MacPorts (Julia seems to choose clang).

A build with gcc with the settings you describe above is accepted by Julia's make system. However, it soon fails with:
```
libtool: compile:  gcc -m64 -DPACKAGE_NAME=\"libuv\" -DPACKAGE_TARNAME=\"libuv\" -DPACKAGE_VERSION=\"0.11.26\" "-DPACKAGE_STRING=\"libuv 0.11.26\"" -DPACKAGE_BUGREPORT=\"https://github.com/joyent/libuv/issues\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libuv\" -DVERSION=\"0.11.26\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_LIBDL=1 -DHAVE_LIBPTHREAD=1 -DHAVE_DTRACE=1 -I. -I/Users/eschnett/src/julia05/deps/srccache/libuv-07730c4bd595b4d45a498a8ee0bcd53878ff7c10 -I/Users/eschnett/src/julia05/deps/srccache/libuv-07730c4bd595b4d45a498a8ee0bcd53878ff7c10/include -I/Users/eschnett/src/julia05/deps/srccache/libuv-07730c4bd595b4d45a498a8ee0bcd53878ff7c10/src -I/Users/eschnett/src/julia05/deps/srccache/libuv-07730c4bd595b4d45a498a8ee0bcd53878ff7c10/src/unix -g -O2 -D_DARWIN_USE_64_BIT_INODE=1 -g -O2 -MT src/unix/libuv_la-async.lo -MD -MP -MF src/unix/.deps/libuv_la-async.Tpo -c /Users/eschnett/src/julia05/deps/srccache/libuv-07730c4bd595b4d45a498a8ee0bcd53878ff7c10/src/unix/async.c  -fno-common -DPIC -o src/unix/.libs/libuv_la-async.o
In file included from /usr/include/Availability.h:172:0,
                 from /usr/include/sys/stat.h:75,
                 from /Users/eschnett/src/julia05/deps/srccache/libuv-07730c4bd595b4d45a498a8ee0bcd53878ff7c10/include/uv-unix.h:26,
                 from /Users/eschnett/src/julia05/deps/srccache/libuv-07730c4bd595b4d45a498a8ee0bcd53878ff7c10/include/uv.h:64,
                 from /Users/eschnett/src/julia05/deps/srccache/libuv-07730c4bd595b4d45a498a8ee0bcd53878ff7c10/src/unix/async.c:25:
/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Headers/LSInfoDeprecated.h:1242:34: error: expected ',' or '}' before '__attribute__'
   kLSHandlerOptionsDefault       __OSX_AVAILABLE_BUT_DEPRECATED_MSG(__MAC_10_4, __MAC_10_11, __IPHONE_4_0, __IPHONE_9_0, "Creator codes are deprecated on OS X.") = 0,
                                  ^
/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Headers/LSOpenDeprecated.h:27:33: error: expected ',' or '}' before '__attribute__'
   kLSLaunchStartClassic         __OSX_AVAILABLE_BUT_DEPRECATED_MSG(__MAC_10_0, __MAC_10_11, __IPHONE_NA, __IPHONE_NA, "The Classic environment is no longer supported.") = 0x00020000, /* Does nothing.*/
                                 ^
make[3]: *** [src/unix/libuv_la-async.lo] Error 1
make[2]: *** [all] Error 2
make[1]: *** [build/libuv-07730c4bd595b4d45a498a8ee0bcd53878ff7c10/.libs/libuv.la] Error 2
make: *** [julia-deps] Error 2
```

That's a known incompatibility between Darwin's system header files and newer versions of gcc. The system header files use a non-standard extension of the C standard (they use `__attribute__` for enums).

--- a/src/julia_threads.h
+++ b/src/julia_threads.h
@@ -52,7 +52,8 @@ typedef struct _jl_tls_states_t {
     jl_jmp_buf *safe_restore;
     int16_t tid;
     size_t bt_size;
-    uintptr_t bt_data[JL_MAX_BT_SIZE + 1];
+    // JL_MAX_BT_SIZE + 1 elements long
+    uintptr_t *bt_data;
 } jl_tls_states_t;
 
 #ifdef __MIC__

--- a/src/threading.c
+++ b/src/threading.c
@@ -105,6 +105,13 @@ static void ti_initthread(int16_t tid)
     ptls->pgcstack = NULL;
     ptls->gc_state = 0; // GC unsafe
     ptls->current_module = NULL;
+    void *bt_data = malloc(sizeof(uintptr_t) * (JL_MAX_BT_SIZE + 1));
+    if (bt_data == NULL) {
+        jl_printf(JL_STDERR, "could not allocate backtrace buffer\n");
+        gc_debug_critical_error();
+        abort();
+    }
+    ptls->bt_data = (uintptr_t*)bt_data;
 #ifdef JULIA_ENABLE_THREADING
     jl_all_heaps[tid] = jl_mk_thread_heap();
 #else

